### PR TITLE
Convert Thelen2003Muscle properties into private subcomponents

### LINKED ISSUE #687: Bugs in properties with Thelen2003Muscle (optimal_fiber_length and activation_time_constant)
Two critical bugs:

1. The `optimal_fiber_length` property exists in both Muscle and in MuscleFixedWidthPennationModel, so the Thelen2003Muscle has this property twice. Somehow, an old model gets the correct value in both places (when the model is updated).
2. The `activation_time_constant` and `deactivation_time_constant` properties were moved from Thelen2003Muscle to MuscleFirstOrderActivationDynamicModel *but are not copied over when updating the model file*. Thus, the updated model just has the default time constants.

Before:

```xml
				<Thelen2003Muscle name="hamstrings_r">
					<!--Flag indicating whether the force is disabled or not. Disabled means that the force is not active in subsequent dynamics realizations.-->
					<isDisabled>false</isDisabled>
					<!--Minimum allowed value for control signal. Used primarily when solving for control values.-->
					<min_control>0</min_control>
					<!--Maximum allowed value for control signal. Used primarily when solving for control values.-->
					<max_control>1</max_control>
					<!--The set of points defining the path of the muscle.-->
					<GeometryPath>
						<!--The set of points defining the path-->
						<PathPointSet>
							<objects>
								<PathPoint name="bifemlh_r-P1">
									<location> -0.129056 -0.105091 0.0711468</location>
									<body>pelvis</body>
								</PathPoint>
								<PathPoint name="bifemlh_r-P2">
									<location> -0.0297545 -0.0355868 0.0290922</location>
									<body>tibia_r</body>
								</PathPoint>
								<PathPoint name="bifemlh_r-P3">
									<location> -0.0231314 -0.0556538 0.0339063</location>
									<body>tibia_r</body>
								</PathPoint>
							</objects>
							<groups />
						</PathPointSet>
						<!--The wrap objecs that are associated with this path-->
						<PathWrapSet>
							<objects />
							<groups />
						</PathWrapSet>
						<!--Used to display the path in the 3D window-->
						<VisibleObject name="display">
							<!--Set of geometry files and associated attributes, allow .vtp, .stl, .obj-->
							<GeometrySet>
								<objects />
								<groups />
							</GeometrySet>
							<!--Three scale factors for display purposes: scaleX scaleY scaleZ-->
							<scale_factors> 1 1 1</scale_factors>
							<!--transform relative to owner specified as 3 rotations (rad) followed by 3 translations rX rY rZ tx ty tz-->
							<transform> -0 0 -0 0 0 0</transform>
							<!--Whether to show a coordinate frame-->
							<show_axes>false</show_axes>
							<!--Display Pref. 0:Hide 1:Wire 3:Flat 4:Shaded Can be overriden for individual geometries-->
							<display_preference>4</display_preference>
						</VisibleObject>
					</GeometryPath>
					<!--The maximum force this actuator can produce.-->
					<optimal_force>1</optimal_force>
					<!--Maximum isometric force that the fibers can generate-->
					<max_isometric_force>2700</max_isometric_force>
					<!--Optimal length of the muscle fibers-->
					<optimal_fiber_length>0.123631055838816</optimal_fiber_length>
					<!--Resting length of the tendon-->
					<tendon_slack_length>0.369758937646367</tendon_slack_length>
					<!--Angle between tendon and fibers at optimal fiber length expressed in radians-->
					<pennation_angle_at_optimal>0</pennation_angle_at_optimal>
					<!--Maximum contraction velocity of the fibers, in optimal fiberlengths/second-->
					<max_contraction_velocity>10</max_contraction_velocity>
					<!--time constant for ramping up muscle activation-->
					<activation_time_constant>0.01</activation_time_constant>
					<!--time constant for ramping down of muscle activation-->
					<deactivation_time_constant>0.04</deactivation_time_constant>
					<!--tendon strain at maximum isometric muscle force-->
					<FmaxTendonStrain>0.033</FmaxTendonStrain>
					<!--passive muscle strain at maximum isometric muscle force-->
					<FmaxMuscleStrain>0.6</FmaxMuscleStrain>
					<!--shape factor for Gaussian active muscle force-length relationship-->
					<KshapeActive>0.5</KshapeActive>
					<!--exponential shape factor for passive force-length relationship-->
					<KshapePassive>4</KshapePassive>
					<!--force-velocity shape factor-->
					<Af>0.3</Af>
					<!--maximum normalized lengthening force-->
					<Flen>1.8</Flen>
				</Thelen2003Muscle>
```

After updating:
```xml
				<Thelen2003Muscle name="hamstrings_r">
					<!--Flag indicating whether the force is disabled or not. Disabled means that the force is not active in subsequent dynamics realizations.-->
					<isDisabled>false</isDisabled>
					<!--Minimum allowed value for control signal. Used primarily when solving for control values.-->
					<min_control>0</min_control>
					<!--Maximum allowed value for control signal. Used primarily when solving for control values.-->
					<max_control>1</max_control>
					<!--The set of points defining the path of the muscle.-->
					<GeometryPath>
						<!--The set of points defining the path-->
						<PathPointSet>
							<objects>
								<PathPoint name="bifemlh_r-P1">
									<location> -0.129056 -0.105091 0.0711468</location>
									<body>pelvis</body>
								</PathPoint>
								<PathPoint name="bifemlh_r-P2">
									<location> -0.0297545 -0.0355868 0.0290922</location>
									<body>tibia_r</body>
								</PathPoint>
								<PathPoint name="bifemlh_r-P3">
									<location> -0.0231314 -0.0556538 0.0339063</location>
									<body>tibia_r</body>
								</PathPoint>
							</objects>
							<groups />
						</PathPointSet>
						<!--The wrap objects that are associated with this path-->
						<PathWrapSet>
							<objects />
							<groups />
						</PathWrapSet>
					</GeometryPath>
					<!--The maximum force this actuator can produce.-->
					<optimal_force>1</optimal_force>
					<!--Maximum isometric force that the fibers can generate-->
					<max_isometric_force>2700</max_isometric_force>
					<!--Optimal length of the muscle fibers-->
					<optimal_fiber_length>0.123631055838816</optimal_fiber_length>
					<!--Resting length of the tendon-->
					<tendon_slack_length>0.369758937646367</tendon_slack_length>
					<!--Angle between tendon and fibers at optimal fiber length expressed in radians-->
					<pennation_angle_at_optimal>0</pennation_angle_at_optimal>
					<!--Maximum contraction velocity of the fibers, in optimal fiberlengths/second-->
					<max_contraction_velocity>10</max_contraction_velocity>
					<!--The model governing the excitation-to-activation dynamics.-->
					<MuscleFirstOrderActivationDynamicModel>
						<!--Activation time constant in seconds-->
						<activation_time_constant>0.015</activation_time_constant>
						<!--Deactivation time constant in seconds-->
						<deactivation_time_constant>0.05</deactivation_time_constant>
					</MuscleFirstOrderActivationDynamicModel>
					<!--The model governing the fiber and tendon kinematics.-->
					<MuscleFixedWidthPennationModel>
						<!--Optimal length of the muscle fibers (overridden when this is a subcomponent of Muscle)-->
						<optimal_fiber_length>0.123631055838816</optimal_fiber_length>
						<!--Angle between tendon and fibers at optimal fiber length expressed in radians (overridden when this is a subcomponent of Muscle)-->
						<pennation_angle_at_optimal>0</pennation_angle_at_optimal>
						<!--Maximum pennation angle (radians)-->
						<maximum_pennation_angle>1.47062890563334</maximum_pennation_angle>
					</MuscleFixedWidthPennationModel>
					<!--tendon strain at maximum isometric muscle force-->
					<FmaxTendonStrain>0.033</FmaxTendonStrain>
					<!--passive muscle strain at maximum isometric muscle force-->
					<FmaxMuscleStrain>0.6</FmaxMuscleStrain>
					<!--shape factor for Gaussian active muscle force-length relationship-->
					<KshapeActive>0.5</KshapeActive>
					<!--exponential shape factor for passive force-length relationship-->
					<KshapePassive>4</KshapePassive>
					<!--force-velocity shape factor-->
					<Af>0.3</Af>
					<!--maximum normalized lengthening force-->
					<Flen>1.8</Flen>
				</Thelen2003Muscle>
```

### PR DESCRIPTION
**Pull request by tkuchida**
Converted `MuscleFixedWidthPennationModel` and `MuscleFirstOrderActivationDynamicModel` properties into private (non-serialized) subcomponents of `Thelen2003Muscle`. The properties of these subcomponents are overridden by Thelen2003Muscle in `extendFinalizeFromProperties()`; temporarily calling `finalizeFromProperties()` on subcomponents manually (see #748). Test cases updated/augmented.

Fixes #687 :rocket:

### CODE REVIEW

File: OpenSim/Actuators/Test/testMuscles.cpp [hunk 1]
```diff
@@ -836,6 +836,97 @@ void testThelen2003Muscle()
...
+            "optimal_fiber_length was not set in pennation model");
+        ASSERT_EQUAL(pennAngAtOptimal, pennMdl2.get_pennation_angle_at_optimal(),
+            SimTK::SignificantReal, __FILE__, __LINE__,
+            "pennation_angle_at_optimal was not set in pennation model");
+        ASSERT_EQUAL(maximumPennation, pennMdl2.get_maximum_pennation_angle(),
+            SimTK::SignificantReal, __FILE__, __LINE__,
+            "maximum_pennation_angle was not set in pennation model");
+
+        // Check properties of MuscleFirstOrderActivationDynamicModel.
+        const MuscleFirstOrderActivationDynamicModel& actMdl2 =
```
Comment (chrisdembia) [inline OpenSim/Actuators/Test/testMuscles.cpp]: Just FYI you could have also done `const auto& actMdl2 = ...`

File: OpenSim/Actuators/Thelen2003Muscle.cpp [hunk 2]
```diff
@@ -105,11 +99,22 @@ void Thelen2003Muscle::extendFinalizeFromProperties()
...
+    pennMdl.set_maximum_pennation_angle(get_maximum_pennation_angle());
+
+    actMdl.set_activation_time_constant(get_activation_time_constant());
+    actMdl.set_deactivation_time_constant(get_deactivation_time_constant());
+    actMdl.set_minimum_activation(get_minimum_activation());
+
+    // Call finalizeFromProperties() on subcomponents manually. These lines can
+    // be removed once support for subcomponents has been fully implemented in
+    // the Component interface. In particular, finalizeFromProperties() must be
+    // called on all subcomponents, not just subcomponents that are properties.
```
Comment (chrisdembia) [inline OpenSim/Actuators/Thelen2003Muscle.cpp]: Can you put "TODO" in the comment?

File: OpenSim/Tests/testIterators/testIterators.cpp [hunk 3]
```diff
@@ -150,7 +150,11 @@ int main()
         // to recompose existing components, this will need continual updating. For example,
         // Joint's often add PhysicalOffsetFrames to handle what used to be baked in location
         // and orientation offsets.
-        ASSERT(numComponents == 77); 
+
+        // Updated below from 77 to 65 when pennation and activation models in
+        // Thelen2003Muscle were converted into private (non-serialized)
+        // subcomponents (2 subcomponents x 6 muscles).
+        ASSERT(numComponents == 65);
```
Comment (chrisdembia) [inline OpenSim/Tests/testIterators/testIterators.cpp]: Hmm I didn't expect this. I thought we would still want to iterate over these subcomponents; that they were just hidden in XML. I now feel that we should *not* include these subcomponents when iterating, but this causes some inconsistency. I imagine that users would still want to be able to access the state variables from these subcomponents (as `/soleus_r/activation_model/activation` or something like that), which would make the class visible in some scenarios, but not visible in others.

**Comment by chrisdembia** [comment]
Mostly looks good to me. Maybe we should decide on a convention for the member variable names of these non-serialized subcomponents; `pennMdl` or `_pennMdl`, `m_pennMdl`, `sub_pennMdl`, etc.?

**Comment by chrisdembia** [comment]
> Mostly looks good to me. Maybe we should decide on a convention for the member variable names of these non-serialized subcomponents; pennMdl or _pennMdl, m_pennMdl, sub_pennMdl, etc.?

Actually, let's not worry about this now. I'd like to create a macro for creating these non-serialized subcomponents, which might make any prior discussion about naming irrelevant.

File: OpenSim/Tests/testIterators/testIterators.cpp (hunk 3 above)
Comment (tkuchida) [inline OpenSim/Tests/testIterators/testIterators.cpp]: I think this is a design question that falls beyond the scope of this PR. Perhaps a new issue should be created for this discussion?

**Comment by tkuchida** [comment]
Thanks for reviewing, @chrisdembia. I added "TODO" to the comment, as requested.

File: OpenSim/Tests/testIterators/testIterators.cpp (hunk 3 above)
Comment (chrisdembia) [inline OpenSim/Tests/testIterators/testIterators.cpp]: I completely agree; sorry, I should have said that.

**Comment by tkuchida** [comment]
Something went wrong with the merge from master and my attempts at fixing it have made things worse. Deleting this branch and creating a new PR.

**Pull request closed without merging**